mail.go

```diff
@@ -0,0 +1,284 @@
+	return m
+}
+
+func (m *Message) AddTo(addresses ...string) *Message {
+	if m.err != nil {
+		return m
+	}
+
+	m.addRecipientAddresses(addresses...)
+
+	return m
+}
+
+// create a multipart MIME message with text and file attachment
+func (m *Message) SetBody(multipartType string, f *excelize.File) *Message {
+	if m.err != nil {
+		return m
+	}
+
+	var header textproto.MIMEHeader
+
+	// create a new multipart writer
+	writer := multipart.NewWriter(m.body)
+	contentType := "multipart/" + multipartType + ";\n \tboundary=" + writer.Boundary()
+	// add header to main header group
+	m.setHeader("Content-Type", contentType)
+
+	// add text part
+	textBody := "\r\nPlease see attached file with portal accounts.\r\n"
+	header = make(textproto.MIMEHeader)
+	header.Set("Content-Type", "text/plain; charset="+charSet)
+	header.Set("Content-Transfer-Encoding", bodyEncoding)
+	p, err := writer.CreatePart(header)
+	if err != nil {
+		m.err = err
+		return m
+	}
+	_, err = p.Write([]byte(b64.StdEncoding.EncodeToString([]byte(textBody))))
+	if err != nil {
+		m.err = err
+		return m
+	}
+
+	// add file attachment part
+	ft, err := newFile(f)
+	if err != nil {
+		m.err = err
+		return m
+	}
+
+	header = make(textproto.MIMEHeader)
+	value := fmt.Sprintf("%s;\n \tname=%q", ft.mimeType, ft.name)
+	header.Set("Content-Type", value)
+	header.Set("Content-Transfer-Encoding", bodyEncoding)
+	value = fmt.Sprintf("attachment;\n \tfilename=%q", ft.name)
+	header.Set("Content-Disposition", value)
+	p, err = writer.CreatePart(header)
+	if err != nil {
+		m.err = err
+		return m
+	}
+	_, err = p.Write([]byte(b64.StdEncoding.EncodeToString(ft.data)))
+	if err != nil {
+		m.err = err
+		return m
+	}
+
+	err = writer.Close()
+	if err != nil {
+		m.err = err
+		return m
+	}
+
+	return m
+}
+
+func (m *Message) MIMEString() string {
+	return m.getHeaders() + m.body.String()
+}
+
+func (m *Message) getHeaders() (headers string) {
+	m.setHeader("Date", time.Now().Format(time.RFC1123Z))
+
+	// combine the headers
+	for header, values := range m.headers {
+		headers += header + ": " + strings.Join(values, ", ") + "\r\n"
+	}
+
+	headers = headers + "\r\n"
+
+	return
+
+}
+
+func (m *Message) validateRecipientAddresses() *Message {
+	// remove duplicate addresses
+	// log invalid address and continue
+	// if no valid address, set m.err
+	addresses := make([]string, 0)
+	addressSet := make(map[string]bool)
+	for _, addr := range m.toAddresses {
+		if len(addr) > 0 {
+			a, err := mail.ParseAddress(addr)
+			if err != nil {
+				log.Printf("validating recipient address %s: %s", addr, err)
+				continue
+			}
+			if _, ok := addressSet[a.Address]; !ok {
+				addressSet[a.Address] = true
+				addresses = append(addresses, a.Address)
+			}
+		} else {
+			continue
+		}
+	}
+	if len(addresses) == 0 {
+		m.err = fmt.Errorf("no valid recipient address")
+		return m
+	}
+
+	m.toAddresses = addresses
+
+	return m
+}
+
+func sendEmail(f *excelize.File) error {
+	smtpRelay := &SMTPRelay{
```

Comment:
this struct also doesn't seem to serve much purpose, since it just "passes" data from the beginning of the function to the end of the same function. I don't see any advantage over just having two variables.

Reply:
Agreed. Removed this code.